### PR TITLE
specs: cleanup/update bedrock derivation specs

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

Update the derivation specs to reflect the latest state of L2 block derivation in Bedrock, cleaning up several TODOs and outdated parts.

- Completed derivation pipeline overview
- Clarify/expand batch submission section to help answer questions that Seb raised.
- Update Engine Queue section to describe the different types of stage updates; forkchoice, consolidation, l1-sync and unsafe payloads.
- Expanded the derivation pipeline reset section with a description of the minimized reset implementation we now have, in favor of the inconsistent/outdated TODO reset sections.
- Replace outdated sync-start pseudo-code with description of current approach
  - Yes, there are minor known issues in this algorithm, but at least it's consistent with the implementation.
  - It's described in a way where it's still clearly separated from the Engine queue reset.
- Moved reorg related note into Derivation pipeline reset sub-section because of the relation to reorgs. And clarified the meaning of finalized L2 data.
- Update wording such as the pulling/pushing of data to reflect the derivation pipeline design adjustments we made over the last half year or so.
- Fixed several broken link-references and typos

**Metadata**

Fix ENG-3141

### CODE REVIEW

**Comment by tynes** [comment]
@refcell Do you mind reading this and commenting on anything that is unclear?

**Comment by refcell** [comment]
> @refcell Do you mind reading this and commenting on anything that is unclear?

Happily!

**Review by refcell — CHANGES_REQUESTED** [review]
Just minor nit literals - otherwise digestible 👍

File: specs/derivation.md [hunk 1]
```diff
@@ -437,16 +422,15 @@ Our architecture decomposes the derivation process into a pipeline made up of th
```
Comment (tynes) [inline specs/derivation.md:422]: nit: prefer numbered lists always using `1.` and then the markdown renderer will auto increment them, this reduces diff when changing orders or adding new items

Comment (protolambda) [reply specs/derivation.md:422]: Agreed, but I've been told not to do this by someone else working on specs before, can't remember who. Shall I just change it to `1.` everywhere in a separate PR, so we never have to touch the numbering again?

**Review by tynes — APPROVED** [review]

File: specs/derivation.md (hunk 1 above)
Comment (tynes) [reply specs/derivation.md:422]: Sure, sorry for the overhead

**Review by refcell — APPROVED** [review]
LGTM 👍

**Comment by protolambda** [comment]
Added do-not-merge until @trianglesphere has taken a look

File: specs/derivation.md [hunk 2]
```diff
@@ -649,32 +636,83 @@ The system configuration is updated with L1 log events whenever the L1 epoch ref
 In the *Engine Queue* stage, the previously derived `PayloadAttributes` structures are buffered and sent to the
 [execution engine][g-exec-engine] to be executed and converted into a proper L2 block.
 
-The engine queue maintains references to two L2 blocks:
+The stage maintains references to three L2 blocks:
 
+- The [finalized L2 head][g-finalized]: everything up to and including this block can be fully derived from the
```
Comment (trianglesphere) [inline specs/derivation.md:641]: nit: Link is not rendering. Maybe g-finalized is off? I'm not exactly sure.

**Review by trianglesphere — COMMENTED** [review]
one broken link, otherwise LGTM

File: specs/derivation.md (hunk 2 above)
Comment (protolambda) [reply specs/derivation.md]: nice catch, fixed.

**Pull request merged**